### PR TITLE
WT-7085 Migrate prepare commit and rollback to use the new hs cursor

### PR DESCRIPTION
**Pull request by quchenhao**

### CODE REVIEW

**Comment by quchenhao** [comment]
@bvpvamsikrishna @hbokhari The test failure is because there is a system issue for macos. Other than that, the tests look fine.

File: src/txn/txn.c [hunk 1]
```diff
@@ -721,24 +721,21 @@ __wt_txn_release(WT_SESSION_IMPL *session)
...
 static int
-__txn_append_hs_record(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, WT_ITEM *key, WT_PAGE *page,
+__txn_append_hs_record(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, WT_PAGE *page,
   WT_UPDATE *chain, bool commit, WT_UPDATE **fix_updp, bool *upd_appended)
 {
-    WT_CURSOR_BTREE *hs_cbt;
     WT_DECL_ITEM(hs_key);
     WT_DECL_ITEM(hs_value);
     WT_DECL_RET;
+    WT_TIME_WINDOW *tw;
```
Comment (hbokhari) [inline src/txn/txn.c:730]: `hs_tw` will be a better name - makes the intention clear.

File: src/txn/txn.c [hunk 2]
```diff
@@ -1128,22 +1089,15 @@ __txn_resolve_prepared_op(WT_SESSION_IMPL *session, WT_TXN_OP *op, bool commit,
...
-        WT_ERR_NOTFOUND_OK(__wt_hs_cursor_position(
-                             session, hs_cursor, hs_btree_id, &op->u.op_row.key, WT_TS_MAX, NULL),
-          true);
-
-        if (ret == 0)
-            /* Not found if we cross the tree or key boundary. */
-            WT_ERR_NOTFOUND_OK(__txn_append_hs_record(session, hs_cursor, &op->u.op_row.key,
-                                 cbt->ref->page, upd, commit, &fix_upd, &upd_appended),
-              true);
+        hs_cursor->set_key(hs_cursor, 4, hs_btree_id, &op->u.op_row.key, WT_TS_MAX, UINT64_MAX);
```
Comment (hbokhari) [inline src/txn/txn.c:1099]: Do we really need to set the counter here? Setting the timestamp as WT_TS_MAX should point us to the latest value anyway.

**Review by hbokhari — COMMENTED** [review]
I have left couple of comments.

File: src/txn/txn.c [hunk 3]
```diff
@@ -748,49 +745,6 @@ __txn_append_hs_record(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, WT_ITEM *
     WT_ERR(__wt_scr_alloc(session, 0, &hs_key));
```
Comment (bvpvamsikrishna) [inline src/txn/txn.c:745]: hs_key is not used any where, so no need of it.

File: src/txn/txn.c [hunk 4]
```diff
@@ -804,10 +758,11 @@ __txn_append_hs_record(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, WT_ITEM *
     if (hs_stop_durable_ts != WT_TS_MAX && commit)
```
Comment (bvpvamsikrishna) [inline src/txn/txn.c:758]: Please have an elaborate comment to explain how this could be possible.

File: src/txn/txn.c [hunk 5]
```diff
@@ -982,33 +939,37 @@ __txn_fixup_prepared_update(
...
      */
-    WT_ERR(__wt_upd_alloc_tombstone(session, &hs_upd, NULL));
     if (commit) {
-        hs_upd->start_ts = txn->commit_timestamp;
-        hs_upd->durable_ts = txn->durable_timestamp;
-        hs_upd->txnid = txn->id;
+        tw.stop_ts = txn->commit_timestamp;
+        tw.durable_stop_ts = txn->durable_timestamp;
+        tw.stop_txn = txn->id;
+        tw.start_ts = fix_upd->start_ts;
```
Comment (bvpvamsikrishna) [inline src/txn/txn.c:946]: better to use WT_TIME_WINDOW_SET_START

File: src/txn/txn.c [hunk 6]
```diff
@@ -1212,15 +1169,14 @@ __txn_resolve_prepared_op(WT_SESSION_IMPL *session, WT_TXN_OP *op, bool commit,
...
      * prepared updates are written to the data store. When the page is read back into memory, there
-     * will be only one uncommitted prepared update. There can be a false positive of fixing history
-     * store when handling prepared inserts, but it doesn't cost much.
+     * will be only one uncommitted prepared update.
      */
     if (fix_upd != NULL)
         WT_ERR(__txn_fixup_prepared_update(session, hs_cursor, fix_upd, commit));
 
 err:
     if (hs_cursor != NULL)
```
Comment (bvpvamsikrishna) [inline src/txn/txn.c:1178]: can we move this hs_cursor != NULL inside hs cursor close function.

File: src/txn/txn.c (hunk 2 above)
Comment (quchenhao) [reply src/txn/txn.c:1099]: Yes, we do. Otherwise counter is set to 0 by default.

File: src/txn/txn.c (hunk 6 above)
Comment (quchenhao) [reply src/txn/txn.c:1178]: I don't think that is possible because close is a cursor call on the cursor pointer itself.

File: src/txn/txn.c (hunk 2 above)
Comment (quchenhao) [reply src/txn/txn.c:1093]: And suppose we have multiple values with the same timestamp, we are not guaranteed to place on the latest one.

File: src/cursor/cur_hs.c [hunk 7]
```diff
@@ -528,6 +528,21 @@ __curhs_next_visible(WT_SESSION_IMPL *session, WT_CURSOR_HS *hs_cursor)
     return (ret);
 }
 
+/*
+ * __wt_hs_upd_time_window --
+ *     Get the underlying time window of the hs cursor.
+ */
+void
+__wt_hs_upd_time_window(WT_CURSOR *hs_cursor, WT_TIME_WINDOW **twp)
```
Comment (bvpvamsikrishna) [inline src/cursor/cur_hs.c:536]: Could we move this to other history store file.

**Review by bvpvamsikrishna — APPROVED** [review]
lgtm, when moved __wt_hs_upd_time_window  to other history store file.

File: src/cursor/cur_hs.c (hunk 7 above)
Comment (quchenhao) [reply src/cursor/cur_hs.c:536]: Which one you are thinking of? I think this file should be all the hs api live.

Comment (bvpvamsikrishna) [reply src/cursor/cur_hs.c:536]: I am thinking of either hs_cursor.c or hs_rec.c

Comment (quchenhao) [reply src/cursor/cur_hs.c:536]: OK, I'll move it to hs_cursor.c

**Review by hbokhari — APPROVED** [review]
LGTM

**Pull request merged**